File: languages/e/elixir/lambda-core.exs
#!/usr/bin/env elixir

###########
#  LOGIC  #
###########

_true = fn x -> fn y -> x end end
_false = fn x -> fn y -> y end end

_not = fn b -> b.(_false).(_true) end
_and = fn b1 -> fn b2 -> b1.(b2).(_false) end end
_or = fn b1 -> fn b2 -> b1.(_true).(b2) end end

#####################
#  CHURCH NUMERALS  #
#####################

_zero = fn f -> fn x -> x end end
_succ = fn n -> fn f -> fn x -> f.(n.(f).(x)) end end end

_pred = fn n ->
  fn f ->
    fn x ->
      n.(fn g -> fn h -> h.(g.(f)) end end).(fn u -> x end).(fn a -> a end)
    end
  end
end

_one = _succ.(_zero)

########################################
#  HELPERS - not pure lambda calculus  #
########################################

read_bool = fn b -> IO.puts(b.("t").("f")) end
read_church = fn n -> IO.puts(n.(fn x -> x + 1 end).(0)) end

##############
#  EXAMPLES  #
##############

IO.puts("LOGIC")
IO.puts("--------------")
IO.puts("TRUE/FALSE")
# t
read_bool.(_true)
# f
read_bool.(_false)

IO.puts("NOT")
# f
read_bool.(_not.(_true))
# t
read_bool.(_not.(_false))

IO.puts("AND")
# t
read_bool.(_and.(_true).(_true))
# f
read_bool.(_and.(_true).(_false))
# f
read_bool.(_and.(_false).(_true))
# f
read_bool.(_and.(_false).(_false))

IO.puts("OR")
# t
read_bool.(_or.(_true).(_true))
# t
read_bool.(_or.(_true).(_false))
# t
read_bool.(_or.(_false).(_true))
# f
read_bool.(_or.(_false).(_false))

IO.puts("\nCHURCH NUMERALS")
IO.puts("--------------")
IO.puts("ZERO/SUCC")
# 0
read_church.(_zero)
# 1
read_church.(_succ.(_zero))
# 2
read_church.(_succ.(_succ.(_zero)))
# 3
read_church.(_succ.(_succ.(_succ.(_zero))))

IO.puts("PRED")
# 0
read_church.(_pred.(_one))
# 1
read_church.(_pred.(_succ.(_one)))
# 2
read_church.(_pred.(_succ.(_succ.(_one))))

File: languages/j/javascript/julia/lambda-core.jl
# LOGIC
const _true  = x -> y -> x
const _false = x -> y -> y
const _not   = b -> b(_false)(_true)
const _and   = b1 -> b2 -> b1(b2)(_false)
const _or    = b1 -> b2 -> b1(_true)(b2)

# CHURCH NUMERALS
const _zero = f -> x -> x
const _succ = n -> f -> x -> f(n(f)(x))
const _pred = n -> f -> x -> n(g -> h -> h(g(f)))(_u -> x)(_a -> a)
const _one  = _succ(_zero)

# HELPERS - not pure lambda calculus
function readBool(b)
    println(b("t")("f"))
end

function readChurch(n)
    println(n(x -> x+1)(0))
end

# -------------------------------------------------

# EXAMPLES
println("LOGIC")
println("---------------")
println("TRUE/FALSE")
readBool(_true)    # t
readBool(_false)   # f

println("NOT")
readBool(_not(_true))   # f
readBool(_not(_false))  # t

println("AND")
readBool(_and(_true)(_true))     # t
readBool(_and(_true)(_false))    # f
readBool(_and(_false)(_true))    # f
readBool(_and(_false)(_false))   # f

println("OR")
readBool(_or(_true)(_true))      # t
readBool(_or(_true)(_false))     # t
readBool(_or(_false)(_true))     # t
readBool(_or(_false)(_false))    # f

println("\nCHURCH NUMERALS")
println("---------------")
println("ZERO/SUCC")
readChurch(_zero)                     # 0
readChurch(_one)                      # 1
readChurch(_succ(_one))               # 2
readChurch(_succ(_succ(_one)))        # 3

println("PRED")
readChurch(_pred(_one))               # 0
readChurch(_pred(_succ(_one)))        # 1
readChurch(_pred(_succ(_succ(_one)))) # 2